src/tribler/gui/widgets/torrentfiletreewidget.py

```diff
@@ -277,4 +289,626 @@ def update_selected_files_size(self, item, _):
+        else:
+            # Scrolling up
+            self.truncate_pages(last_visible_page + self.view_size_post)
+            for page_index in range(self.previous_view_start_index,
+                                    max(0, self.view_start_index - self.view_size_pre),
+                                    -1):
+                # Reload any unloaded pages!
+                if not self.pages[page_index].loaded:
+                    self.fetch_page(page_index)
+                self.reset_scroll_bar()
+
+        # Hard refresh all visible rows, in case of really fast scrolling and tearing. This is important on slower
+        # machines, which can "outscroll" the Qt updates.
+        self.refresh_visible()
+
+    def on_expanded_contracted(self, response) -> None:
+        """
+        The core finished expanding or collapsing a directory.
+
+        ALL pages after and including the page that the expansion/collapse happened need to be refreshed.
+        """
+        if response is None:
+            return
+
+        page_index = self.exp_contr_requests.pop(response["path"])
+        if page_index is None:
+            return
+        page_index = max(0, page_index - 1)
+
+        self.truncate_pages(page_index)
+
+        request_manager.get(
+            endpoint=f"downloads/{self.infohash}/files",
+            url_params={
+                'view_start_path': self.pages[page_index].query,
+                'view_size': self.page_size
+            },
+            on_success=self.fill_entries,
+        )
+
+    def hideEvent(self, a0) -> None:
+        """
+        We are not shown, no need to do the loading animation.
+        """
+        super().hideEvent(a0)
+        self.loading_movie.stop()
+
+    def showEvent(self, a0) -> None:
+        """
+        We are shown, continue the loading animation.
+        """
+        super().showEvent(a0)
+        self.loading_movie.start()
+
+    def resizeEvent(self, e) -> None:
+        """
+        We got resized, causing the previous first and last visible row to be invalidated: perform a hard refresh.
+        """
+        super().resizeEvent(e)
+        if self.isVisible():
+            self.refresh_visible()
+
+    def refresh_visible(self) -> None:
+        """
+        Determine the visible rows and refresh what we are missing.
+
+        Note that fetch_page will drop unattainable pages beyond our current knowledge and fill_entries will recursively
+        pull those pages in afterward.
+        """
+        first_visible_row = self.rowAt(0)
+        if first_visible_row == -1:
+            first_visible_row = 0
+        last_visible_row = self.rowAt(self.height() - 1)
+        if last_visible_row == -1:
+            last_visible_row = first_visible_row
+        for page_index in (range(self.view_start_index, self.view_start_index + last_visible_row - first_visible_row)
+                           or [self.view_start_index]):
+            self.fetch_page(page_index)
```

Comment:
This is inelegant. It would be nicer to add refresh logic for page ranges that preceed the tail of the `pages` list, just like the incremental growth after the (current) tail of the pages list.